src/link.rs

```diff
@@ -62,13 +62,41 @@ fn collect_roots(
     roots
 }
 
+/// Ensure that all `CrateIndex`es have consistent schema versions. If all
+/// indexes have the same schema version, return the version number. Otherwise,
+/// return an error.
+///
+/// Precondition: there must be at least one `CrateIndex`. If this precondition
+/// is not met, this function will panic.
+fn check_equal_versions(
+    indexes: &[CrateIndex],
+) -> serde_cbor::Result<u64> {
+    match indexes.split_first() {
+        Some((first_index, last_indexes)) => {
+            let version = first_index.version;
+            for index in last_indexes {
+                if version != index.version {
+                    return Err(serde::ser::Error::custom(format!(
+                        "Mismatched schema version numbers ({:?} and {:?})",
+                        version,
+                        index.version,
+                    )));
+                }
+            }
+            Ok(version)
+        },
+        None =>
+            panic!("check_equal_versions: Empty list of crate indexes"),
+    }
+}
 
 /// Combine the contents of `ocs`, producing a combined JSON crate data object as the result.
 pub fn link_crates<R, W>(inputs: &mut [R], mut output: W) -> serde_cbor::Result<()>
 where R: Read + Seek, W: Write {
     let (indexes, json_offsets) = read_crates(inputs)?;
     let (it, defs, translate) = assign_global_ids(&indexes);
     let roots = collect_roots(&indexes, &translate);
+    let version = check_equal_versions(&indexes)?;
```

Comment:
It might be worth being stricter here and asserting that `version` matches the current `SCHEMA_VER`, or at least checking that it's new enough to have the same set of `EntryKind`s (and other top-level structure, such as `roots`).  Currently, if we remove an `EntryKind` in schema v2, trying to link together a bunch of v1 inputs using the v2 tools will produce an invalid "v1" output that only has tables for the new `EntryKind`s and omits the removed one.

Reply:
Good point—it doesn't seem terribly useful to have `mir-json` (which is assumed to work with only one schema version) be able to link files that were produced with an older schema. I've reworked this function to check for this as well.